File: src/puppy_core/puppy_core/capability_registry.py
"""Capability Registry: declare and query system capabilities."""
from dataclasses import dataclass
from typing import Dict, Set


@dataclass
class Capability:
    """A single system capability."""

    name: str
    enabled: bool = True
    available: bool = True
    reason: str = ""


class CapabilityRegistry:
    """Registry of system capabilities for graceful degradation."""

    def __init__(self):
        self._caps: Dict[str, Capability] = {}
        self.register('navigation')
        self.register('patrol')
        self.register('security')
        self.register('emotion')
        self.register('fall_detection')
        self.register('docking')
        self.register('rgb_camera')
        self.register('depth_camera')
        self.register('lidar')
        self.register('imu')

    def register(self, name: str, enabled: bool = True, available: bool = True):
        self._caps[name] = Capability(name, enabled, available)

    def is_available(self, name: str) -> bool:
        cap = self._caps.get(name)
        return cap is not None and cap.enabled and cap.available

    def set_enabled(self, name: str, enabled: bool, reason: str = ""):
        if name in self._caps:
            self._caps[name].enabled = enabled
            if not enabled:
                self._caps[name].reason = reason or 'disabled by config'
            elif self._caps[name].available:
                self._caps[name].reason = ""

    def disable(self, name: str, reason: str = ""):
        if name in self._caps:
            self._caps[name].available = False
            self._caps[name].reason = reason

    def enable(self, name: str):
        if name in self._caps and self._caps[name].enabled:
            self._caps[name].available = True
            self._caps[name].reason = ""

    def get_status(self) -> Dict[str, dict]:
        return {
            name: {'enabled': cap.enabled, 'available': cap.available, 'reason': cap.reason}
            for name, cap in self._caps.items()
        }

    def get_available(self) -> Set[str]:
        return {name for name, cap in self._caps.items() if cap.enabled and cap.available}

File: cpp_src/puppy_core/capability_registry.cpp
// Capability Registry 实现
//
// 纯工具类（非 ROS2 节点），对应 Python: puppy_core/capability_registry.py
#include "puppy_core/capability_registry.h"

namespace puppy_core {
namespace registry {

CapabilityRegistry::CapabilityRegistry() {
  registerCapability("navigation");
  registerCapability("patrol");
  registerCapability("security");
  registerCapability("emotion");
  registerCapability("fall_detection");
  registerCapability("docking");
  registerCapability("rgb_camera");
  registerCapability("depth_camera");
  registerCapability("lidar");
  registerCapability("imu");
}

void CapabilityRegistry::registerCapability(const std::string& name,
                                            bool enabled, bool available) {
  Capability cap;
  cap.name = name;
  cap.enabled = enabled;
  cap.available = available;
  caps_[name] = cap;
}

bool CapabilityRegistry::isAvailable(const std::string& name) const {
  auto it = caps_.find(name);
  if (it == caps_.end()) {
    return false;
  }
  return it->second.enabled && it->second.available;
}

void CapabilityRegistry::setEnabled(const std::string& name, bool enabled,
                                    const std::string& reason) {
  auto it = caps_.find(name);
  if (it == caps_.end()) {
    return;
  }
  it->second.enabled = enabled;
  if (!enabled) {
    it->second.reason = reason.empty() ? "disabled by config" : reason;
  } else if (it->second.available) {
    it->second.reason = "";
  }
}

void CapabilityRegistry::disable(const std::string& name,
                                 const std::string& reason) {
  auto it = caps_.find(name);
  if (it == caps_.end()) {
    return;
  }
  it->second.available = false;
  it->second.reason = reason;
}

void CapabilityRegistry::enable(const std::string& name) {
  auto it = caps_.find(name);
  if (it == caps_.end()) {
    return;
  }
  if (it->second.enabled) {
    it->second.available = true;
    it->second.reason = "";
  }
}

std::map<std::string, std::map<std::string, std::string>>
CapabilityRegistry::getStatus() const {
  std::map<std::string, std::map<std::string, std::string>> out;
  for (const auto& kv : caps_) {
    const Capability& c = kv.second;
    auto& entry = out[kv.first];
    entry["enabled"] = c.enabled ? "true" : "false";
    entry["available"] = c.available ? "true" : "false";
    entry["reason"] = c.reason;
  }
  return out;
}

std::set<std::string> CapabilityRegistry::getAvailable() const {
  std::set<std::string> out;
  for (const auto& kv : caps_) {
    if (kv.second.enabled && kv.second.available) {
      out.insert(kv.first);
    }
  }
  return out;
}

}  // namespace registry
}  // namespace puppy_core
